app/models/profile.rb

```diff
@@ -25,9 +25,9 @@ def decrypt_pii(user_access_key)
   end
 
   def recover_pii(recovery_code)
-    raise 'invalid recovery code' unless recovery_code_generator.verify(recovery_code)
-    uak = recovery_code_generator.user_access_key
-    Pii::Attributes.new_from_encrypted(encrypted_pii_recovery, uak)
+    rc_user_access_key = UserAccessKey.new(recovery_code, ssn_signature)
```

Comment:
It's really salt-per-User and salt-per-Profile.

The dilemma is UX. We really want the user to have a minimal number of things to remember, and so we are dual-purposing the recovery code for (a) authenticating the User and (b) decrypting the recovery PII.

In one sense it would be better to have `RecoveryCodeGenerator` operate on a `User` or a `Profile` and manage their salts and recovery codes separately. But that would introduce a 2nd recovery code for the user to write down. Maybe that's not such a bad idea from a data modeling perspective, but it does put a lot of burden on the user: they can't simply use whatever recovery code they happen to have to do both (a) and (b).

You're right about adding a salt specific to the Profile that isn't the SSN. Hopefully the SSN eventually becomes as optional to the app as it is in the model.